### PR TITLE
accounts/abi: implement abi.encodePacked

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
This PR implements the `abi.EncodePacked` function a non-standard encoding function available in solidity.
Doc: https://docs.soliditylang.org/en/latest/abi-spec.html#non-standard-packed-mode
Some behaviour around arrays are guesstimates based on evaluations did with solidity.

rel: https://github.com/ethereum/go-ethereum/issues/22257

### CODE REVIEW

**Comment by axic** [comment]
Pardon my ignorance, but what use cases will this be used for? I think the long term goal is to replace that legacy feature with some much nicer bytestream reader/writer. There are some discussions about various options to accomplish it.

**Comment by MariusVanDerWijden** [comment]
@axic I wasn't aware that this should be replaced. My idea was to add this to abi, s.th. it is possible for people to `encodePacked` similar to what all web3.js, ethers,... provide. This is part of a longer project we have to enable users to easily interact with their smart contracts using go.
Could you point me to the discussions?

**Comment by axic** [comment]
Here is a longer discussion with various options: https://github.com/ethereum/solidity/issues/9829

The likely outcome will be to have something like `concat`, fixed to variable length bytes conversion, and slicing.

The main use case of `abi.encodePacked` is for contract to be able to assemble data for hashing. It is not for exchanging data. It's support for arrays and various types is patchy and has not been touched for a long while.

**Comment by MariusVanDerWijden** [comment]
Use case would be to locally hash the data in the same case to verify signatures or create signatures that can be verified by the contract

**Comment by axic** [comment]
I was not aware ethers-js had it implemented (somehow the call it a hashing algorithm?): https://docs.ethers.io/v5/api/utils/hashing/#utils--solidity-hashing

It is not well documented (just the general difference between the actual ABI and this) so not sure what exactly is implemented.

File: accounts/abi/encodepacked.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+	"math/big"
+	"reflect"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// EncodePacked implements the non-standard encoding available in solidity.
+// Since encoding is ambigious there is no decoding function.
+// See
+func EncodePacked(args []Type, values []interface{}) ([]byte, error) {
```
Comment (axic) [inline accounts/abi/encodepacked.go:16]: If this is merged I would suggest to go with `SolidityEncodePacked` to highlight the fact this is some Solidity specific feature.

**Comment by ekpyron** [comment]
I'm not sure if I'm missing something there, but to me it would seem that using a non-injective encoding for the purpose of hashing data is a pretty bad choice (your hash function can be as good as it gets, if you collapse distinct data by encoding it to the same byte string...) - so if it was up to me, I'd strongly discourage anyone from using ``abi.encodePacked`` like that (not that it might not still make sense for geth to support it, if it is for whatever reason used like that in practice, but maybe it'd be good to at least mention this somewhere? We should probably also emphasize that in the solidity docs, if it isn't there already... If it was less of an issue to break the solidity ABI, I would think that we'd most probably have switched away from ``abi.encodePacked`` for events a long time ago as well...). The only type signatures for which this is a sane choice is a list of value types followed by at most one array... and even for those cases I'm not sure there's that much advantage compared to other encodings that work injectively for all kinds of data...

**Comment by axic** [comment]
@MariusVanDerWijden thank your for bring this up. Here are some more thoughts on it: https://github.com/ethereum/solidity/issues/10903

**Comment by MariusVanDerWijden** [comment]
Pinging @holiman, @ricmoo and @cgewecke

As Daniel said it is not safe to use `abi.EncodePacked` in hashing as for example `abi.EncodePacked(uint8(0xFF), uint8(0xFF))` produces the same encoding as `abi.EncodePacked(uint16(0xFFFF))` which could be used to trick users into signing bogus stuff. As Alex said, most users use `abi.EncodePacked` because they want to `concat` two datatypes.

Solidity already has a disclaimers that this is unsafe. It would be good imo for ethers.js and web3.js to add similar disclaimers to their `ethers.utils.solidityKeccak256` and `web3.utils.soliditySha3` documenation.

**Comment by ricmoo** [comment]
The [examples given](https://docs.ethers.io/v5/api/utils/hashing/#utils--solidity-hashing) are meant to highlight the problem with the non-standard, and using the language non-standard is meant to deter people from using it. Ethers must provide this though, since it is something people have used historically and they may need to pass in a signed payload (or some other internal use) of the hashed data.

I can add more to the docs indicating it is not generally safe, but whenever issues come up, I make sure to warn people away from it. :)

**Comment by ricmoo** [comment]
From my docs:

```javascript
// As a short example of the non-distinguished nature of
// Solidity tight-packing (which is why it is inappropriate
// for many things from a security point of view), consider
// the following examples are all equal, despite representing
// very different values and layouts.

utils.solidityPack([ "string", "string" ], [ "hello", "world01" ])
// '0x68656c6c6f776f726c643031'

utils.solidityPack([ "string", "string" ], [ "helloworld", "01" ])
// '0x68656c6c6f776f726c643031'

utils.solidityPack([ "string", "string", "uint16" ], [ "hell", "oworld", 0x3031 ])
// '0x68656c6c6f776f726c643031'

utils.solidityPack([ "uint96" ], [ "32309054545061485574011236401" ])
// '0x68656c6c6f776f726c643031'
```

**Comment by cgewecke** [comment]
cc: @GregTheGreek

**Comment by fjl** [comment]
We should add some test cases from ethers.js to make sure this implementation is correct

**Comment by ricmoo** [comment]
Quick note, the test cases in ethers were created using [procedurally generated](https://github.com/ethers-io/ethers.js/blob/v4-legacy/tests/make-tests/make-contract-events.js#L578) Solidity source, which were then compiled, deployed and executed in Geth/Parity.

The full tests are available as a portable [gzipped json](https://github.com/ethers-io/ethers.js/blob/master/packages/testcases/testcases/solidity-hashes.json.gz) and is an array of [these objects](https://docs.ethers.io/v5/testing/#testing--testing-schemas--solidity-hashes).

Hope that helps. :)

File: accounts/abi/encodepacked.go (hunk 1 above)
Comment (Haoxiqiang) [reply accounts/abi/encodepacked.go]: > 
ethers-js impl: `ethers.utils.solidityPack`

**Pull request closed without merging**